torch_dftd/nn/base_dftd_module.py

```diff
@@ -145,19 +138,38 @@ def calc_energy_and_forces(
                 results_list[i]["forces"] = forces[batch == i].cpu().numpy()
 
         if cell is not None:
-
             # stress = torch.mm(cell_grad, cell.T) / cell_volume
             # Get stress in Voigt notation (xx, yy, zz, yz, xz, xy)
+            assert isinstance(shift, Tensor)
+            voigt_left = [0, 1, 2, 1, 2, 0]
+            voigt_right = [0, 1, 2, 2, 0, 1]
             if batch is None:
                 cell_volume = torch.det(cell).abs()
-                stress = torch.mm(cell.grad, cell.T) / cell_volume
-                stress = stress.view(-1)[[0, 4, 8, 5, 2, 1]]
+                cell_grad = torch.sum(
+                    (pos[:, voigt_left] * pos.grad[:, voigt_right]).to(torch.float64), dim=0
+                )
+                cell_grad += torch.sum(
+                    (shift[:, voigt_left] * shift.grad[:, voigt_right]).to(torch.float64), dim=0
+                )
```

Comment:
Step-by-step instruction for this line:

The previous implementation is `cell.grad`. The cell was used in two places. One is `pos` and another is distance calculation combined with `shift`. Therefore, we can write down the calculation of `cell.grad` by explicitly write the backpropagation of `cell`->`pos` and `cell`->`shift`. The above two sentences corresponds to them.

I will use the character `*` for matrix multiplication here. For `pos` side, the forward calculation is
`pos = rel_pos * cell`,

therefore the backward calculation is
`cell.grad = rel_pos.T * pos.grad`.

Since `cell.T` will be multiplied from left later (line 153 in the previous implementation, see corresponding comment), the calculation will be
`cell.T * cell.grad = cell.T * rel_pos.T * pos.grad = pos.T * pos.grad`.

Here, the shape of `pos` and `g_pos` are both `(n_node, 3)`, the matrix calculation is `(3, n_node) * (n_node, 3) -> (3, 3)` ans it can also be written explicitly like
`pos.T * pos.grad = torch.sum(pos[:, :, None] * pos.grad[:, None, :], dim=0)`.

The obtained value is stress tensor and from the physical consideration we know it is always symmetry. Therefore we just pick up diagonal and upper triangular values (Voigt notation). Seeing the above calculation, we can modify it so that only needed values will be calculated.
`stress = (pos.T * pos.grad).view(-1)[[0, 4, 8, 5, 2, 1]] = torch.sum(pos[:, [0,1,2,1,2,0]] * pos.grad[:, [0,1,2,2,0,1]], dim=0)`

This is exactly what the new implementation does. Since the summed axis length is `n_node`, we use FP64 to avoid numerical calculation error.

Almost same discussion can be applied to the relation between `cell.grad` and `shift.grad`.